File: ggcommons/metrics/targets/messaging.py
from awsiot.greengrasscoreipc.model import QOS
from ggcommons.messaging.message_builder import MessageBuilder
from ggcommons.messaging.messaging_client import MessagingClient
from ggcommons.config.manager.config_manager import ConfigManager
from ggcommons.metrics.targets.emf_helper import build_metric_data_emf
from ggcommons.metrics.targets.metric_target import MetricTarget


def _is_local_destination(destination: str) -> bool:
    """True for the local/IPC transport, False for IoT Core.

    Canonical values are "ipc" (local/IPC) and "iot_core" (IoT Core); the legacy
    "local"/"iotcore" spellings are also accepted. Mirrors the heartbeat target,
    the Java/Rust metric targets, and the config schema.
    """
    return destination.lower() in ("ipc", "local")


class Messaging(MetricTarget):
    def __init__(self, config_manager: ConfigManager):
        super().__init__(config_manager)
        self.topic = config_manager.resolve_template(
            config_manager.get_metric_config().get_topic()
        )
        self.send_to_local = _is_local_destination(
            config_manager.get_metric_config().get_destination()
        )

    def emit_metric_now(self, metric, measure_values):
        metric_name = metric.get_name()
        self.logger.debug(f"Emitting metric '{metric_name}' to messaging target with {len(measure_values)} measures")
        
        metric_data = build_metric_data_emf(
            self.metric_config, metric, measure_values, False
        )
        self.__publish_message(metric_data)
        
        if self.metric_config.get_large_fleet_workaround():
            self.logger.debug(f"Emitting large fleet workaround metric for '{metric_name}'")
            metric_data = build_metric_data_emf(
                self.metric_config, metric, measure_values, True
            )
            self.__publish_message(metric_data)

        self.logger.debug(f"Metric '{metric_name}' emission completed")

    def __publish_message(self, metric_dict: dict):
        destination = "local" if self.send_to_local else "IoT Core"
        self.logger.debug(f"Publishing metric message to {destination} on topic: {self.topic}")
        
        message = MessageBuilder.create("Metric", "1.0") \
            .with_payload(metric_dict) \
            .with_config(self.config_manager) \
            .build()
            
        if self.send_to_local:
            MessagingClient.publish(self.topic, message)
        else:
            MessagingClient.publish_to_iot_core(self.topic, message, QOS.AT_LEAST_ONCE)

    def on_configuration_change(self, configuration) -> bool:
        self.logger.info("Metric messaging configuration changed, reconfiguring target")
        
        old_topic = self.topic
        old_destination = "local" if self.send_to_local else "IoT Core"
        
        self.topic = self.config_manager.resolve_template(
            self.config_manager.get_metric_config().get_topic()
        )
        self.send_to_local = _is_local_destination(
            self.config_manager.get_metric_config().get_destination()
        )

        new_destination = "local" if self.send_to_local else "IoT Core"
        
        self.logger.info(f"Metric messaging reconfigured - topic: {old_topic} -> {self.topic}, destination: {old_destination} -> {new_destination}")
        return True

File: libs/java/src/main/java/com/mbreissi/ggcommons/metrics/targets/Messaging.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.ggcommons.metrics.targets;

import com.mbreissi.ggcommons.config.ConfigManager;
import com.mbreissi.ggcommons.messaging.Message;
import com.mbreissi.ggcommons.messaging.MessageBuilder;
import com.mbreissi.ggcommons.messaging.MessageIdentity;
import com.mbreissi.ggcommons.messaging.MessagingClient;
import com.mbreissi.ggcommons.metrics.Metric;
import com.mbreissi.ggcommons.uns.Uns;
import com.mbreissi.ggcommons.uns.UnsClass;
import com.google.gson.JsonObject;
import software.amazon.awssdk.aws.greengrass.model.QOS;

import java.util.Map;

/**
 * The {@code messaging} metric target (UNS-CANONICAL-DESIGN §4.3): publishes each metric to the
 * library-owned UNS metric topic {@code ecv1/{device}/{component}/main/metric/{metricName}} (the
 * metric name sanitized as a channel token) through the privileged
 * {@link com.mbreissi.ggcommons.messaging.ReservedPublisher} seam — the {@code metric} class is
 * reserved. {@code metricEmission.targetConfig.destination} still selects local/IPC vs IoT Core
 * (D-U9); the legacy {@code targetConfig.topic} override is removed.
 */
public final class Messaging extends MetricTarget {

    private boolean sendToIpc = true;
    private MessagingClient messagingService;
    /** WARN-once flag for the no-resolved-identity (test/subclass bring-up) case. */
    private boolean warnedNoIdentity = false;

    public Messaging(ConfigManager configManager) {
        super(configManager);
        this.sendToIpc = !isIotCoreDestination(metricConfig.getDestination());
    }

    /**
     * IoT Core is selected only by "iot_core"/"iotcore"; everything else (the
     * canonical "ipc", the legacy "local", and any unrecognized value) uses the
     * local/IPC transport, so a metric never routes to a possibly-unconfigured
     * IoT Core. Matches the heartbeat destination and the Python/Rust metric targets.
     */
    private static boolean isIotCoreDestination(String destination) {
        return destination.equalsIgnoreCase("iot_core") || destination.equalsIgnoreCase("iotcore");
    }

    public void setMessagingService(MessagingClient messagingService) {
        this.messagingService = messagingService;
    }

    @Override
    public void emitMetricNow(Metric metric, Map<String, Float> measureValues)
    {
        JsonObject metricObject = EmfHelper.buildMetricData(metricConfig.getNamespace(), metric, measureValues, false);
        publishMessage(metric, metricObject);

        if (metricConfig.getLargeFleetWorkaround())
        {
            metricObject = EmfHelper.buildMetricData(metricConfig.getNamespace(), metric, measureValues, true);
            publishMessage(metric, metricObject);
        }
    }

    /**
     * The metric's UNS topic — {@code ecv1[/{site}]/{device}/{component}/main/metric/{name}} with
     * the metric name passed through the template sanitizer (the §2.2 channel-token rule), or
     * {@code null} (WARN once) when no component identity is resolved (mock/test bring-up).
     */
    private String metricTopic(Metric metric)
    {
        MessageIdentity identity = configManager.getComponentIdentity();
        if (identity == null)
        {
            if (!warnedNoIdentity)
            {
                warnedNoIdentity = true;
                LOGGER.warn("No resolved component identity - the messaging metric target cannot"
                        + " build UNS metric topics; metrics are dropped");
            }
            return null;
        }
        return new Uns(identity, configManager.isTopicIncludeRoot())
                .topic(UnsClass.METRIC, ConfigManager.sanitize(metric.getName()));
    }

    private void publishMessage(Metric metric, JsonObject metricObject)
    {
        String topic = metricTopic(metric);
        if (topic == null)
        {
            return;
        }
        Message message = MessageBuilder.create("Metric", "1.0")
                .withPayload(metricObject)
                .withConfig(configManager)
                .build();
        // The metric class is reserved (§4.1) - publish through the privileged seam (§4.2).
        if (sendToIpc)
            messagingService.reservedPublisher().publish(topic, message);
        else
            messagingService.reservedPublisher().publishToIoTCore(topic, message, QOS.AT_LEAST_ONCE);
        LOGGER.trace("Metric emitted for {} emitted", metric);
    }

    @Override
    public boolean onConfigurationChanged()
    {
        LOGGER.info("Configuration changed. Reconfiguring metric messaging destination");
        this.sendToIpc = !isIotCoreDestination(configManager.getMetricConfig().getDestination());
        return true;
    }

    @Override
    public void emitMetric(Metric metric, Map<String, Float> measureValues)
    {
        emitMetricNow(metric, measureValues);
    }
}
